### PR TITLE
feat(docker): Provides ubi8 based images for che-theia and che-theia-dev

### PR DESCRIPTION
**Pull request by benoitf**
### What does this PR do?
Provides ubi8 image for che-theia and che-theia-dev

Pure ubi8 (no external repository needed)
- disable some electron dependencies as some 3rd libraries are not provided in default ubi8 and that we don't rely on it for che-theia (use only browser part, not electron)
- use same yarn version than in alpine

Provides a system of includes and overrides

example:

`${INCLUDE:instructions/${BUILD_IMAGE_TARGET}/runtime-install-dependencies.run}`
will include a file named `instructions/rhel/runtime-install-dependencies.run` for `rhel` based image and `instructions/default/runtime-install-dependencies.run` for Alpine/default image

so downstream components can tweak files easily

it's integrated with existing build system:

`./build.sh` will build two images (alpine + rhel) (rhel will be suffixed to image name)
`./build.sh  --dockerfile:Dockerfile.rhel` will build rhel image to `eclipse/che-theia:next`

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/13765

#### Release Notes
N/A

#### Docs PR
N/A

Change-Id: Ia29d9ecdc61b9d4e02c5b205d7591a5424fb7cc7
Signed-off-by: Florent Benoit <fbenoit@redhat.com>

### CODE REVIEW

**Comment by dmytro-ndp** [comment]
crw-ci-test

File: dockerfiles/theia-dev/build-scripts/default/01-install-dependencies.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+apk add --update --no-cache \
+  curl `# Download some files` \
+  jq `# To play with json in shell` \
+  make gcc g++ python `# compile some javascript native stuff (node-gyp)` \
+  git openssh openssh-keygen `# clone repositories (and also using ssh repositories)` \
+  less `# Handle git diff properly` \
+  bash `# bash shell` \
+  libx11-dev libxkbfile-dev `# some lib to compile 'native-keymap' npm mpdule`
+
```
Comment (nickboldt) [inline dockerfiles/theia-dev/build-scripts/default/01-install-dependencies.sh]: do we need to install yarn here, like in https://github.com/eclipse/che-theia/pull/467/files#diff-b601fdde7ca9d45a0d2f10471ca7bfdf ?

Comment (benoitf) [reply dockerfiles/theia-dev/build-scripts/default/01-install-dependencies.sh]: no we don't, yarn is already installed on community images this is why I haven't installed it

Comment (benoitf) [reply dockerfiles/theia-dev/build-scripts/default/01-install-dependencies.sh]: https://github.com/nodejs/docker-node/blob/a9c583095d4cf08bbd68f570a1f9a99780820351/10/alpine/Dockerfile#L61

**Comment by nickboldt** [comment]
crw-ci-test

File: dockerfiles/build.include [hunk 2]
```diff
@@ -138,7 +142,35 @@ build() {
 build_image() {
   printf "${BOLD}Building Docker Image ${IMAGE_NAME} from $DIR directory with tag $TAG${NC}\n"
   # Replace macros in Dockerfiles
-  cat ${DIR}/${DOCKERFILE} | sed s/\$\{BUILD_ORGANIZATION\}/${ORGANIZATION}/ | sed s/\$\{BUILD_PREFIX\}/${PREFIX}/ | sed s/\$\{BUILD_TAG\}/${TAG}/ | sed s/\$\{BUILD_PARENT_IMAGE\}/${PARENT_IMAGE}/ | sed s/\$\{GIT_BRANCH_NAME\}/${BRANCH}/ | sed s/\$\{GIT_REF\}/"${GIT_REF}"/ > ${DIR}/.Dockerfile
+  cat ${DIR}/${DOCKERFILE} | sed s/\$\{BUILD_ORGANIZATION\}/${ORGANIZATION}/ | sed s/\$\{BUILD_PREFIX\}/${PREFIX}/ \
```
Comment (nickboldt) [inline dockerfiles/build.include]: You do this exact same bunch of string replacements on line 166. Why do it twice?

File: dockerfiles/build.include [hunk 3]
```diff
@@ -138,7 +142,35 @@ build() {
...
+    fi
+  done
+  cp ${DIR}/.Dockerfile ${DIR}/.Dockerfile2
+  cat ${DIR}/.Dockerfile2 | sed s/\$\{BUILD_ORGANIZATION\}/${ORGANIZATION}/ | sed s/\$\{BUILD_PREFIX\}/${PREFIX}/ \
+  | sed s/\$\{BUILD_TAG\}/${TAG}/ | sed s/\$\{BUILD_PARENT_IMAGE\}/${PARENT_IMAGE}/ \
+  | sed s/\$\{BUILD_IMAGE_TARGET\}/${BUILD_IMAGE_TARGET}/  \
+  | sed s/\$\{GIT_BRANCH_NAME\}/${BRANCH}/ | sed s/\$\{GIT_REF\}/"${GIT_REF}"/ > ${DIR}/.Dockerfile
+  rm ${DIR}/.Dockerfile2
+
+
```
Comment (nickboldt) [inline dockerfiles/build.include]: couple extra empty lines here we don't need.

File: dockerfiles/theia/instructions/runtime-configure.run [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    && echo '{"optOut": true}' > $HOME/.config/insight-nodejs/insight-yo.json \
+    # Link yarn global modules for yeoman
+    && local_modules=$(ls -d1 /usr/*/node_modules 2>/dev/null || ls -d1 /usr/*/*/node_modules) \
+    && mv ${local_modules}/* /usr/local/share/.config/yarn/global/node_modules && rm -rf ${local_modules} && ln -s /usr/local/share/.config/yarn/global/node_modules $(dirname ${local_modules})/ \
+    # Cleanup tmp folder
+    && rm -rf /tmp/* \
+    # Cleanup yarn cache
+    && yarn cache clean \
+    # Change permissions to allow editing of files for openshift user
+    && find ${HOME} -exec sh -c "chgrp 0 {}; chmod g+rwX {}" \;
```
Comment (nickboldt) [inline dockerfiles/theia/instructions/runtime-configure.run]: I think we already did this step on line 10. Why do it twice? Or why do it earlier if you have to do it again? Maybe we omit ${HOME} on line 10?

**Comment by benoitf** [comment]
@nickboldt what are these commits on my PR ?
Please avoid to hijack my PRs

**Comment by nickboldt** [comment]
I was making it work for Brew :)

**Comment by benoitf** [comment]
closing PR as I don't endorsed changes

**Comment by nickboldt** [comment]
would you prefer a PR against your branch?

**Comment by nickboldt** [comment]
well, I removed the commits from your branch ...

https://github.com/eclipse/che-theia/tree/CHE-13765

... but seems this PR is broken now. Oops.

Anyway I like what you have here - it's working for an online RHEL 8 build.

But to make it work in Brew, I need these additional changes:

https://github.com/nickboldt/che-theia/tree/CHE-13765

**Pull request closed without merging**